### DISCUSSION TITLE
Is there any way to programatically add page objects before any files are processed?

### DISCUSSION THREAD

**Discussion by Simon-Martens** [11ty/eleventy#2713] [Q&A] [answered]
Say, I have a huge XML file and I wanted to take a set of elements of the file to programmatically create separated pages. I would take care of selecting the portions, generating permalinks & titles (all that stuff that is done with front matter usually). Is there any way to "inject" these kinds of self generated objects, so 11ty creates pages from them.

I've looked at `custom file formats`, which is kinda the wrong thing to do here, since it's about parsing of templating languages and not adding whole pages and it doesn't allow for manipulating front matter. 

Then, there would be the option to preprocess & dump all the parts as part of a huge data file and use pagination to generate pages for that. Is this te only way to go about this? Or can I generate page-like-objects and insert them manually into the data cascade?

**Comment by pdehaan** [comment]
Not sure if I understand 100%, but if the question is: "Can i use XML as a data file format?", see https://www.11ty.dev/docs/data-custom/.
If you want to dynamically create pages from data files, I'd use pagination. https://www.11ty.dev/docs/pagination/#paginate-a-global-or-local-data-file

**Comment by Simon-Martens** [comment, marked as answer]
I got it to work. Read in my xml file, manually created a collection and added the parts of the xml that I wanted to be on different pages as nodes into the collection. See here: https://www.11ty.dev/docs/collections/#advanced-custom-filtering-and-sorting

Then I used the pagination feature and paginated the collection using a size of 1. See: https://www.11ty.dev/docs/pagination/#paging-a-collection

I still would prefer if one could create a ton of objects like:
```
{
  "data": 
  {
    "url": "/x/y/z/",
    "title": "test title",
    "date": "xxxx-yy-zz"
  },
  "content": "xyz"
}
```

And add them through a programmatic API as if it was a file on disc. Since there would be no input path, one had to declare the permalink manually. 

Anyway it works right now, thanks :) have a great new year
